app/src/main/java/net/christianbeier/droidvnc_ng/InputService.java

```diff
@@ -12,302 +12,291 @@
  *
  * Swipe fixes and gesture handling by Christian Beier <info@christianbeier.net>.
  *
+ * Expanded hotkey system Alexander Balanyuk Anatolevich QuantMad@gmail.com
+ *
  */
 
 import android.accessibilityservice.AccessibilityService;
 import android.accessibilityservice.GestureDescription;
 import android.content.ClipData;
 import android.content.ClipboardManager;
 import android.content.Context;
+import android.graphics.Path;
 import android.os.Handler;
 import android.util.DisplayMetrics;
 import android.util.Log;
+import android.view.ViewConfiguration;
 import android.view.WindowManager;
 import android.view.accessibility.AccessibilityEvent;
-import android.view.ViewConfiguration;
-import android.graphics.Path;
+
+import java.util.Arrays;
 
 public class InputService extends AccessibilityService {
 
-	/**
-	 * This globally tracks gesture completion status and is _not_ per gesture.
-	 */
-	private static class GestureCallback extends AccessibilityService.GestureResultCallback {
-		private boolean mCompleted = true; // initially true so we can actually dispatch something
-
-		@Override
-		public synchronized void onCompleted(GestureDescription gestureDescription) {
-			mCompleted = true;
-		}
-
-		@Override
-		public synchronized void onCancelled(GestureDescription gestureDescription) {
-			mCompleted = true;
-		}
-	}
-
-	private static final String TAG = "InputService";
-
-	private static InputService instance;
-
-	private Handler mMainHandler;
-
-	private boolean mIsButtonOneDown;
-	private Path mPath;
-	private long mLastGestureStartTime;
-
-	private boolean mIsKeyCtrlDown;
-	private boolean mIsKeyAltDown;
-	private boolean mIsKeyShiftDown;
-	private boolean mIsKeyDelDown;
-	private boolean mIsKeyEscDown;
-
-	private float mScaling;
-
-	private final GestureCallback mGestureCallback = new GestureCallback();
-
-
-	@Override
-	public void onAccessibilityEvent( AccessibilityEvent event ) { }
-
-	@Override
-	public void onInterrupt() { }
-
-	@Override
-	public void onServiceConnected()
-	{
-		super.onServiceConnected();
-		instance = this;
-		mMainHandler = new Handler(instance.getMainLooper());
-		Log.i(TAG, "onServiceConnected");
-	}
-
-	@Override
-	public void onDestroy() {
-		super.onDestroy();
-		instance = null;
-		Log.i(TAG, "onDestroy");
-	}
-
-	public static boolean isEnabled()
-	{
-		return instance != null;
-	}
-
-	/**
-	 * Set scaling factor that's applied to incoming pointer events by dividing coordinates by
-	 * the given factor.
-	 * @param scaling The scaling factor as a real number.
-	 * @return Whether scaling was applied or not.
-	 */
-	public static boolean setScaling(float scaling) {
-		try {
-			instance.mScaling = scaling;
-			return true;
-		} catch (Exception e) {
-			return false;
-		}
-	}
-
-	@SuppressWarnings("unused")
-	public static void onPointerEvent(int buttonMask, int x, int y, long client) {
-
-		try {
-			x /= instance.mScaling;
-			y /= instance.mScaling;
+    /**
```

Comment:
Please make sure to not indent code that's not being changed. This makes reviewing changes very hard :-)

Reply:
Oh. I'm really sorry. I've never made a pull request before. This seems a bit tricky. I can pull the latest version, make my edits where needed, and repeat the pull-request. Will it save the situation?

Reply:
> 

Apologies in advance for my English. I rely a lot on Google translator.
I don't know if my approach is architecturally correct, but here's what it is:
There is a VNCKey enumeration, each member of which corresponds to a keysym, and has one parameter that is responsible for the current state of the key (pressed / released). The onKeyEvent event of the InputService passes the corresponding key (if it was defined in the enum) a new status. Thus, at any point in the program, when accessing the button, you can find out its current state.
Hotkey processing also has a fairly simple principle: There is a small "HotkeyBinding" class, which includes an array of three VNCKeys, and a stub for anonymous classes implemented on the basis of the ICommand interface. This interface has only one method: void execute();, which is responsible for what exactly this keyboard shortcut does. Further, the matter is small: in the InputService, you can add any key combination to the static constant HOTKEY_BINDINGS by writing a couple of lines.

As for the user interface, the code looks scary, but it works. All shortcuts are added to the user interface dynamically, and do not require any intervention in the process. The programmer just needs to add a new line to the HOTKEY_BINDINGS constant, and he does not have to worry about everything else. That was the idea